src/fileoverview_comment_transformer.ts

```diff
@@ -61,89 +61,106 @@ function augmentFileoverviewComments(tags: jsdoc.Tag[]) {
  * A transformer that ensures the emitted JS file has an \@fileoverview comment that contains an
  * \@suppress {checkTypes} annotation by either adding or updating an existing comment.
  */
-export function transformFileoverviewComment(context: ts.TransformationContext):
-    (sf: ts.SourceFile) => ts.SourceFile {
-  return (sf: ts.SourceFile) => {
-    const text = sf.getFullText();
-
-    let fileComments: ts.SynthesizedComment[] = [];
-    const firstStatement = sf.statements.length && sf.statements[0] || null;
-
-    const originalComments = ts.getLeadingCommentRanges(text, 0) || [];
-    if (!firstStatement) {
-      // In an empty source file, all comments are file-level comments.
-      fileComments = synthesizeCommentRanges(sf, originalComments);
-    } else {
-      // Search for the last comment split from the file with a \n\n. All comments before that are
-      // considered fileoverview comments, all comments after that belong to the next statement(s).
-      // If none found, comments remains empty, and the code below will insert a new fileoverview
-      // comment.
-      for (let i = originalComments.length - 1; i >= 0; i--) {
-        const end = originalComments[i].end;
-        if (!text.substring(end).startsWith('\n\n')) continue;
-        // This comment is separated from the source file with a double break, marking it (and any
-        // preceding comments) as a file-level comment. Split them off and attach them onto a
-        // NotEmittedStatement, so that they do not get lost later on.
-        const synthesizedComments = jsdoc.synthesizeLeadingComments(firstStatement);
-        const notEmitted = ts.createNotEmittedStatement(sf);
-        // Modify the comments on the firstStatement in place by removing the file-level comments.
-        fileComments = synthesizedComments.splice(0, i + 1);
-        // Move the fileComments onto notEmitted.
-        ts.setSyntheticLeadingComments(notEmitted, fileComments);
-        // TODO(martinprobst): consider checking here whether any of the trailing comments contains
-        // an @fileoverview etc, and reporting an error if so.
-        sf = updateSourceFileNode(
-            sf, ts.createNodeArray([notEmitted, firstStatement, ...sf.statements.slice(1)]));
-        break;
+export function transformFileoverviewCommentFactory(diagnostics: ts.Diagnostic[]) {
+  return (): (sourceFile: ts.SourceFile) => ts.SourceFile => {
+    function checkNoFileoverviewComments(
+        context: ts.Node, comments: jsdoc.SynthesizedCommentWithOriginal[], message: string) {
+      for (let j = 0; j < comments.length; j++) {
```

Comment:
Looks like this could be a for-of, or at least use 'i' as index.

Reply:
Done.